### PR TITLE
replace usages of ``copy_arrays`` with ``memmap`` for ``asdf>=3.1.0``

### PR DESCRIPTION
**Pull request by zacharyburnett**
follow-on to https://github.com/asdf-format/asdf/pull/1797

### CODE REVIEW

**Comment by braingram** [comment]
Hmmm... the oldestdeps should have failed here since `asdf==2.13` (our minimum) doesn't have memmap:
https://github.com/astropy/asdf-astropy/blob/dab5b4d73ce3f75e60e5b29e75eb3e811dfd8319/pyproject.toml#L21
It doesn't seem to even be running `minimum_dependencies` :-/ and is picking up asdf 3.3.0
https://github.com/astropy/asdf-astropy/actions/runs/9993708226/job/27621947749?pr=230#step:10:24

@zacharyburnett what do you think we should do here? The usage is only in tests so run-time asdf-astropy doesn't need asdf>=3.1.0 but if we fix our oldest deps it will start failing (since it uses the runtime requirements). The tests would also fail for anyone who packages this repository using the run-time requirements to run tests (like the asdf issue that came up yesterday). We could update the tests to check the asdf version (and use either `copy_arrays` or `memmap`) but I'm not sure if it's really worth it (although that would allow us to fix the oldestdeps and hopefully have it not fail).

**Comment by zacharyburnett** [comment]
> Hmmm... the oldestdeps should have failed here since `asdf==2.13` (our minimum) doesn't have memmap:
> 
> https://github.com/astropy/asdf-astropy/blob/dab5b4d73ce3f75e60e5b29e75eb3e811dfd8319/pyproject.toml#L21
> 
> 
> It doesn't seem to even be running `minimum_dependencies` :-/ and is picking up asdf 3.3.0
> https://github.com/astropy/asdf-astropy/actions/runs/9993708226/job/27621947749?pr=230#step:10:24
> @zacharyburnett what do you think we should do here? The usage is only in tests so run-time asdf-astropy doesn't need asdf>=3.1.0 but if we fix our oldest deps it will start failing (since it uses the runtime requirements). The tests would also fail for anyone who packages this repository using the run-time requirements to run tests (like the asdf issue that came up yesterday). We could update the tests to check the asdf version (and use either `copy_arrays` or `memmap`) but I'm not sure if it's really worth it (although that would allow us to fix the oldestdeps and hopefully have it not fail).

I think the only two options we have that maintain green status are
- remove oldestdeps testing or
- use `copy_arrays` or `memmap` based on version

I'm inclined toward the latter.

File: asdf_astropy/converters/unit/tests/test_quantity.py
Comment (braingram) [inline asdf_astropy/converters/unit/tests/test_quantity.py:12]: This package can use the (sometimes confusing) `minversion` from astropy:
https://github.com/astropy/asdf-astropy/blob/dab5b4d73ce3f75e60e5b29e75eb3e811dfd8319/asdf_astropy/converters/transform/core.py#L134

File: asdf_astropy/converters/unit/tests/test_quantity.py
Comment (braingram) [inline asdf_astropy/converters/unit/tests/test_quantity.py:139]:
```suggestion
    with asdf.open(file_path, mode="rw", **asdf_open_memory_mapping_kwarg(memmap=False)) as af:
```

File: asdf_astropy/converters/unit/tests/test_quantity.py
Comment (braingram) [inline asdf_astropy/converters/unit/tests/test_quantity.py:149]:
```suggestion
    with asdf.open(file_path, mode="rw", **asdf_open_memory_mapping_kwarg(memmap=False)) as af:
```

File: asdf_astropy/converters/unit/tests/test_quantity.py
Comment (braingram) [inline asdf_astropy/converters/unit/tests/test_quantity.py:14]:
```suggestion
    else:
```

File: asdf_astropy/converters/unit/tests/test_quantity.py
Comment (braingram) [inline asdf_astropy/converters/unit/tests/test_quantity.py:11]: The `bool` annotation here is leading to a pre-commit/ruff failure:
https://results.pre-commit.ci/run/github/271820376/1721652625.YDD78eseT3m3vTW-_PD9kw
https://docs.astral.sh/ruff/rules/boolean-type-hint-positional-argument/
I am ok with either of the following:
- a global ignore for all `FBT` rules.
- removing this type hint

**Comment by braingram** [comment]
@zacharyburnett would you rebase this to (hopefully) allow the oldestdeps to use the oldest versions (now that https://github.com/astropy/asdf-astropy/pull/232 is in).

Also pre-commit is complaining of a `RET505` error.
https://results.pre-commit.ci/run/github/271820376/1721656013.8blo7lrzSCa9pRAmDI-kog

File: asdf_astropy/converters/unit/tests/test_quantity.py
Comment (braingram) [inline asdf_astropy/converters/unit/tests/test_quantity.py:15]:
```suggestion
    return {"copy_arrays": not memmap}
```
pre-commit is picky

**Review by braingram — APPROVED** [review]
Thanks!

**Comment by braingram** [comment]
Oh, looks like we don't have branch protection rules 😮  Let's wait for the CI to go green before merging.

**Pull request merged**
